File: learning-game-engine/zig/src/ui/components/button.zig
const c = @import("../../platform/sdl.zig").c;
const Input = @import("../../input/input.zig").Input;
const text = @import("text.zig");

pub const Button = struct {
    label: [*c]const u8,
    rect: c.SDL_FRect,
    active: bool,
    on_click: *const fn (ctx: *anyopaque) void,
    ctx: *anyopaque,
};

fn inside(r: c.SDL_FRect, x: f32, y: f32) bool {
    return x >= r.x and x <= r.x + r.w and y >= r.y and y <= r.y + r.h;
}

pub fn button(b: Button, input: *const Input, renderer: *c.SDL_Renderer, font: ?*c.TTF_Font, scale: f32) void {
    var r = b.rect;
    if (b.active) {
        _ = c.SDL_SetRenderDrawColor(renderer, 77, 166, 242, 255);
    } else {
        _ = c.SDL_SetRenderDrawColor(renderer, 50, 50, 58, 255);
    }
    _ = c.SDL_RenderFillRect(renderer, &r);

    const text_color = if (b.active) c.SDL_Color{ .r = 0, .g = 0, .b = 0, .a = 255 } else c.SDL_Color{ .r = 230, .g = 230, .b = 230, .a = 255 };
    text.drawTextCentered(renderer, font, b.label, b.rect, text_color, scale);

    if (input.mouse_clicked and inside(b.rect, input.mouse_x, input.mouse_y)) {
        b.on_click(b.ctx);
    }
}

File: learning-game-engine/rust/src/ui/components/button.rs
use sdl3::pixels::Color;
use sdl3::render::{FRect, WindowCanvas};
use sdl3::ttf::Font;

use super::text::draw_text_centered;

fn inside(r: FRect, x: f32, y: f32) -> bool {
    x >= r.x && x <= r.x + r.w && y >= r.y && y <= r.y + r.h
}

pub fn button(
    canvas: &mut WindowCanvas,
    font: &Font,
    scale: f32,
    label: &str,
    rect: FRect,
    active: bool,
    mouse_clicked: bool,
    mx: f32,
    my: f32,
) -> bool {
    if active {
        canvas.set_draw_color(Color::RGBA(77, 166, 242, 255));
    } else {
        canvas.set_draw_color(Color::RGBA(50, 50, 58, 255));
    }
    let _ = canvas.fill_rect(rect);

    let text_color = if active {
        Color::RGBA(0, 0, 0, 255)
    } else {
        Color::RGBA(230, 230, 230, 255)
    };
    draw_text_centered(canvas, font, label, rect, text_color, scale);

    mouse_clicked && inside(rect, mx, my)
}
